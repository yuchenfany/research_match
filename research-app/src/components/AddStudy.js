/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import Select from 'react-select';
import { useNavigate } from 'react-router-dom';
import '../assets/index.css';
// study, setStudy,
function AddStudy({ study, setStudy }) {
  // const updateStudy = async () => {
  //   setStudy({
  //     title: '',
  //     description: '',
  //     compensation: '',
  //     duration: '',
  //     tags: [],
  //     participants: [],
  //     studyId: 4,
  //     researchers: [],
  //   });
  // };

  // useEffect(() => {
  //   setStudy({
  //     title: '',
  //     description: '',
  //     compensation: '',
  //     duration: '',
  //     tags: [],
  //     participants: [],
  //     studyId: 4,
  //     researchers: [],
  //   });
  //   console.log('useEFfect is being run');
  //   console.log(study.description);
  // }, []);
  const navigate = useNavigate();
  const Tags = [
    { label: 'Diabetes', value: 'diabetes' },
    { label: 'Cancer', value: 'cancer' },
    { label: 'Social', value: 'social' },
    { label: 'placebo', value: 'placebo' },
    { label: 'Brain', value: 'brain' },
    { label: 'Physical', value: 'physical' },
  ];
  const customStyles = {
    menu: (provided, state) => ({
      ...provided,
      width: 'state.selectProps.width,',
      color: state.selectProps.menuColor,
    }),
    control: (provided, state) => ({
      ...provided,
      background: '#F9FFFE',
      borderColor: '#808A8F',
      fontSize: '14px',
      boxShadow: state.isFocused ? null : null,
    }),
    multiValueLabel: (styles) => ({
      ...styles,
      backgroundColor: '#BBEFEB',
    }),
    multiValueRemove: (styles) => ({
      ...styles,
      backgroundColor: '#BBEFEB',
      ':hover': {
        color: '#4aa8a2',
      },
    }),
  };
  // async function getNextStudyID() {
  //   const studyData = await fetch('http://localhost:5000/findMax', {
  //     method: 'GET',
  //     headers: {
  //       'Content-Type': 'application/json',
  //     },
  //   });
  //   const data = await studyData.json();
  //   console.log(data);
  //   return data;
  // }

  // async function addStudy() {
  //   await fetch('http://localhost:5000/add-study', {
  //     method: 'POST',
  //     headers: {
  //       'Content-Type': 'application/json',
  //     },
  //     body: JSON.stringify(study),
  //   })
  //     .catch((e) => {
  //       window.alert(e);
  //     });

  //   console.log('fetch is being called');
  // }
  async function verify() {
    const studyData = await fetch('http://localhost:5000/findMax', {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    const data = await studyData.json();
    const Id = data[0].studyId + 1;
    console.log(Id);
    // console.log(study.studyId);
    // console.log(nextId);
    setStudy({
      title: study.title,
      description: study.description,
      compensation: study.compensation,
      duration: study.duration,
      tags: study.tags,
      participants: study.participants,
      studyId: Id,
      researchers: study.researchers,
    // }, () => {
    //   addStudy();
    });
    const myobj = {
      title: study.title,
      description: study.description,
      compensation: study.compensation,
      duration: study.duration,
      tags: study.tags,
      participants: study.participants,
      studyId: Id,
      researchers: study.researchers,
    };
    console.log(Id);
    await fetch('http://localhost:5000/add-study', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(myobj),
    })
      .catch((e) => {
        window.alert(e);
      });

    console.log('fetch is being called');
    return true;
  }

  const updateTitle = async (event) => {
    console.log('UPDATE TITLE');
    console.log(event.target.value);
    setStudy({
      title: event.target.value,
      description: study.description,
      compensation: study.compensation,
      duration: study.duration,
      tags: study.tags,
      participants: study.participants,
      studyId: study.studyId,
      researchers: study.researchers,
    });
  };
  const updateDescription = async (event) => {
    console.log('UPDATE description');
    console.log(event.target.value);
    setStudy({
      title: study.title,
      description: event.target.value,
      compensation: study.compensation,
      duration: study.duration,
      tags: study.tags,
      participants: study.participants,
      studyId: study.studyId,
      researchers: study.researchers,
    });
  };
  const updateCompensation = async (event) => {
    setStudy({
      title: study.title,
      description: study.description,
      compensation: event.target.value,
      duration: study.duration,
      tags: study.tags,
      participants: study.participants,
      studyId: study.studyId,
      researchers: study.researchers,
    });
  };
  const updateDuration = async (event) => {
    setStudy({
      title: study.title,
      description: study.description,
      compensation: study.compensation,
      duration: event.target.value,
      tags: study.tags,
      participants: study.participants,
      studyId: study.studyId,
      researchers: study.researchers,
    });
  };
  const updateResearcher = async (event) => {
    setStudy({
      title: study.title,
      description: study.description,
      compensation: study.compensation,
      duration: study.duration,
      tags: study.tags,
      participants: study.participants,
      studyId: study.studyId,
      researchers: event.target.value,
    });
  };
  async function getTagsArr(tags) {
    const arr = [];

    for (let i = 0; i < tags.length; i += 1) {
      arr.push(tags[i].value);
    }

    return arr;
  }

  const updateTags = async (tags) => {
    const arr = await getTagsArr(tags);
    console.log(arr);
    setStudy({
      title: study.title,
      description: study.description,
      compensation: study.compensation,
      duration: study.duration,
      tags: arr,
      participants: study.participants,
      studyId: study.studyId,
      researchers: study.researchers,
    });
  };
  async function handleSubmit(event) {
    if (await verify()) {
      navigate('/home');
    } else {
      event.preventDefault();
    }
  }

  return (
    <div className="Profile">
      <div className="profile-flex">
        <div className="header-left"> Create Study </div>
        <div className="title-row">
          <div>Title</div>
          <input
            className="input-field"
            type="text"
            id="title"
            onChange={updateTitle}
          />
        </div>
        <div className="description-row">
          <div>Description</div>
          <input
            className="input-field"
            type="text"
            id="description"
            onChange={updateDescription}
          />
          <div>Compensation</div>
          <input
            className="input-field"
            type="text"
            id="compensation"
            onChange={updateCompensation}
          />
          <div>Duration</div>
          <input
            className="input-field"
            type="text"
            id="duration"
            onChange={updateDuration}
          />
          <div>Lead Researcher</div>
          <input
            className="input-field"
            type="text"
            id="researchers"
            onChange={updateResearcher}
          />
        </div>
        <div className="profile-row">
          <div>Tags</div>
        </div>
        <div className="profile-row">
          <Select
            options={Tags}
            isMulti
            onChange={(tags) => updateTags(tags)}
            styles={customStyles}
          />
        </div>
        <input className="signup-button" type="submit" value="Add Study" onClick={handleSubmit} />
      </div>
    </div>
  );
}

export default AddStudy;
