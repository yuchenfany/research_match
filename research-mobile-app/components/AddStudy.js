/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */

import React, { useState, useEffect } from 'react';
// import NavBar from './NavBar';

function AddStudy({ route, navigation}) { // add props user
  let {user, setUser, setStudy } = route.params;
  const [enrolledStudies, setEnrolledStudies] = useState([]);
  // const navigate = useNavigate();
  async function getStudyIds() {
    const data = await fetch(`http://localhost:5000/record/${user.username}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    const json = await data.json();
    setUser({
      username: user.username,
      password: user.password,
      name: user.name,
      organization: user.organization,
      studies: json.studies,
      type: user.type,
    });
    // setUser({ username: user.username, password: user.password, enrolled: json.enrolled });
    return json?.studies ?? [];
  }

  // gets individual study by id
  async function getStudy(studyId) {
    const data = await fetch(`http://localhost:5000/study/${studyId}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    console.log(data);
    return data.json();
  }

  // get all studies
  async function getAllStudyJson() {
    const studyIds = await getStudyIds();
    console.log(user.username);
    console.log(user.studies);
    console.log(studyIds);
    return Promise.all(studyIds.map((studyId) => getStudy(studyId)));
  }

  useEffect(() => {
    getAllStudyJson()
      .then(setEnrolledStudies);
  }, []);

  function goToStudy(studyId) {
    setStudy({ studyId });
    navigation.navigate(`/researcher-study/${studyId}`);
  }

  return (
    <div className="ResearcherProfile">
      <div className="header-left">Add Study Page</div>
      <div className="study-flex">
        <div className="header-left">Enrolled Studies</div>
        <div>
          {
          enrolledStudies.length === 0 ? []
            : enrolledStudies.map(
              (studyJson) => (
                <div key={studyJson.studyId} className="study">
                  <div className="study-title">{studyJson.title}</div>
                  <button className="view-button" type="button" key={studyJson.studyId} onClick={() => goToStudy(studyJson.studyId)}>VIEW</button>
                </div>
              ),
            )
          }
        </div>
      </div>
      <div className="study-transfer">
        <div className="header-left">For Testing Purposes: Directs to Add Study Page</div>
        <div className="study">
          <div className="study-transfer">Go to Study Page</div>
          <button className="view-button" type="button" onClick={() => navigation.navigate('ResearcherHome', {user: user, setUser: setUser, setStudy: setStudy})}>Go back to Home</button>
        </div>
      </div>
    </div>
  );
}

export default AddStudy;
