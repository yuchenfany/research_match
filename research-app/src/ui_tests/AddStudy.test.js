/**
 * @jest-environment jsdom
 */

/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import React from 'react';
// import { jest } from '@jest/globals';

import AddStudy from '../components/AddStudy';

// const mockedUsedNavigate = jest.fn();

// jest.mock('react-router-dom', () => ({
//   ...jest.requireActual('react-router-dom'),
//   useNavigate: () => mockedUsedNavigate,
// }));

test('Create Study header', () => {
  render(
    <Router>
      <AddStudy />
    </Router>,
  );
  const header = screen.getByText('Create Study');
  expect(header).toBeInTheDocument();
});

test('Title label', () => {
  render(
    <Router>
      <AddStudy />
    </Router>,
  );
  const label = screen.getByText('Title');
  expect(label).toBeInTheDocument();
});

test('Description label', () => {
  render(
    <Router>
      <AddStudy />
    </Router>,
  );
  const label = screen.getByText('Description');
  expect(label).toBeInTheDocument();
});

test('Compensation label', () => {
  render(
    <Router>
      <AddStudy />
    </Router>,
  );
  const label = screen.getByText('Compensation');
  expect(label).toBeInTheDocument();
});

test('Duration label', () => {
  render(
    <Router>
      <AddStudy />
    </Router>,
  );
  const label = screen.getByText('Duration');
  expect(label).toBeInTheDocument();
});

test('Tags label', () => {
  render(
    <Router>
      <AddStudy />
    </Router>,
  );
  const label = screen.getByText('Tags');
  expect(label).toBeInTheDocument();
});

test('Add Study sequence', async () => {
  const mockStudy = {
    title: 'faketitle', description: 'fakedescription', compensation: 4, duration: 3, leadResearcher: 'fakelead',
  };
  const mockUser = {
    username: '', password: '',
  };
  const mockSetStudy = jest.fn();
  render(
    <Router>
      <AddStudy user={mockUser} study={mockStudy} setStudy={mockSetStudy} />
    </Router>,
  );

  const inputTitle = screen.getByTitle('title');
  fireEvent.change(inputTitle, { target: { value: 'title' } });
  expect(inputTitle.value).toEqual('title');

  const inputDesc = screen.getByTitle('description');
  fireEvent.change(inputDesc, { target: { value: 'description' } });
  expect(inputDesc.value).toEqual('description');

  const inputComp = screen.getByTitle('compensation');
  fireEvent.change(inputComp, { target: { value: 'compensation' } });
  expect(inputComp.value).toEqual('compensation');

  const inputDur = screen.getByTitle('duration');
  fireEvent.change(inputDur, { target: { value: 'duration' } });
  expect(inputDur.value).toEqual('duration');

  await waitFor(() => {
    fireEvent.click(screen.getByTitle('add-study'));
  });
  expect(mockSetStudy).toHaveBeenCalled();
});

// test('should be able to type input', () => {
//   render(
//     <Router>
//       <AddStudy />
//     </Router>,
//   );
//   const input = screen.getByPlaceholderText('title');
//   fireEvent.change(input, { target: { value: 'test input' } });
//   // expect(input.value).toBe('test input');

//   // expect(screen.getByDisplayValue('title')).toHaveAttribute('id', 'title');
// });

/* eslint-disable */
jest.mock('react-select', () => ({ options, value, onChange }) => {
  function handleChange(event) {
    const option = options.find(
      (option) => option.value === event.currentTarget.value,
    );
    onChange(option);
  }

  return (
    <select data-testid="select" value={value} onChange={handleChange}>
      {options.map(({ label, value }) => (
        <option key={value} value={value}>
          {label}
        </option>
      ))}
    </select>
  );
});
/* eslint-enable */

test('Add Study tags', () => {
  const mockStudy = {
    title: 'faketitle', description: 'fakedescription', compensation: 4, duration: 3, leadResearcher: 'fakelead',
  };
  const mockUser = {
    username: '', password: '',
  };
  const mockSetStudy = jest.fn();
  render(
    <Router>
      <AddStudy user={mockUser} study={mockStudy} setStudy={mockSetStudy} />
    </Router>,
  );
  fireEvent.change(screen.getByTestId('select'), {
    target: { value: 'diabetes' },
  });
});

test('snapshot test', () => {
  const component = renderer.create(
    <Router>
      <AddStudy />
    </Router>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
