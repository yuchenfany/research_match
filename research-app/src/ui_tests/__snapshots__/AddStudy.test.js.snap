// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="Profile"
>
  <div
    className="profile-flex"
  >
    <div
      className="header-left"
    >
       Create Study 
    </div>
    <div
      className="title-row"
    >
      <div>
        Title
      </div>
      <input
        className="input-field"
        id="title"
        onChange={[Function]}
        title="title"
        type="text"
      />
    </div>
    <div
      className="description-row"
    >
      <div>
        Description
      </div>
      <input
        className="input-field"
        id="description"
        onChange={[Function]}
        title="description"
        type="text"
      />
      <div>
        Compensation
      </div>
      <input
        className="input-field"
        id="compensation"
        onChange={[Function]}
        title="compensation"
        type="text"
      />
      <div>
        Duration
      </div>
      <input
        className="input-field"
        id="duration"
        onChange={[Function]}
        title="duration"
        type="text"
      />
    </div>
    <div
      className="profile-row"
    >
      <div>
        Tags
      </div>
    </div>
    <div
      className="profile-row"
    >
      <select
        data-testid="select"
        onChange={[Function]}
      >
        <option
          value="diabetes"
        >
          Diabetes
        </option>
        <option
          value="cancer"
        >
          Cancer
        </option>
        <option
          value="social"
        >
          Social
        </option>
        <option
          value="placebo"
        >
          placebo
        </option>
        <option
          value="brain"
        >
          Brain
        </option>
        <option
          value="physical"
        >
          Physical
        </option>
      </select>
    </div>
    <input
      className="signup-button"
      onClick={[Function]}
      title="add-study"
      type="submit"
      value="Add Study"
    />
  </div>
</div>
`;
